---
title: "April Taiari eDNA Survey"
author: "Christopher Kavazos"
date: '`r Sys.Date()`'
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load Packages

The following packages are required to run this analaysis

```{r package, error = FALSE, message = FALSE, warning=FALSE}
library(readxl)
library(openxlsx)
library(splitstackshape)
library(data.table)
library(lubridate)
library(dplyr)
library(tidyverse)
library(taxa)
library(phyloseq)
library(metacoder)
library(dplyr)
library(ggrepel)
library(microeco)
library(file2meco)
library(pheatmap)
library(magrittr)
library(speedyseq)
library(gridExtra)

set.seed(1)
```

## Load data

This analysis requires three datasets. 'meta' is the metadata that is available in metadata.xlsx. The eDNA results provides the taxonmic read count data and TICI health scores and is available in WLJ602811.xlsx.

```{r pressure, echo=TRUE, message = FALSE}
meta <- read_excel("metadata.xlsx")
taxa_data <- read_xlsx("WLJ602811.xlsx", sheet=2)
TICI <- read_xlsx("WLJ602811.xlsx", sheet=4)

#Chain of custody information is also used to run checks
coa <- read_xlsx("WLJ602811.xlsx", sheet=1)
coa <- coa[19:nrow(coa),]
colnames(coa) <- coa[1,]
coa <- data.frame(coa[2:nrow(coa),])
```

# Import into phyloseq

A number of steps are required to get the data formated correctly for us in phyloseq. Three sets of data are required. 1) The sample_data which is found in 'meta'; 2) The taxa read counts found in 'taxa_data' and 3) the taxonomic information which we will obtain from NCBI.

## Sample data
Prepare sample_data for import into phyloseq

```{r sample}
# mutate data.frame into long format by UID
df <- cSplit(meta, "UID", sep = ",", direction = "long")

#Calculate deployment period
df$`Retrieval date` <- as.Date(df$`Retrieval date`, format = "%y/%m/%d")
df$`Deployment date` <- as.Date(df$`Deployment date`, format = "%y/%m/%d")

Retrieval = data.frame(
  date=df$`Retrieval date`,
  time=format(df$`Retrieval time`, "%H:%M")
)

Deployment = data.frame(
  date=df$`Deployment date`,
  time=format(df$`Deployment time`, "%H:%M")
)

Deployment <- as.POSIXct(paste(Deployment$date, Deployment$time), format="%Y-%m-%d %H:%M")
Retrieval <- as.POSIXct(paste(Retrieval$date, Retrieval$time), format="%Y-%m-%d %H:%M")

#Make new column with sample period
df$period <- round(difftime(Retrieval, Deployment, units = "hours"),1)

#Trim data frame to UID's with read counts
df <- df[which(df$UID %in% coa$UID),]
```

Data from REC2 model is incorporated into metadata (https://niwa.co.nz/freshwater-and-estuaries/management-tools/river-environment-classification-0). Cumulative land uses data are to be transformed here to a percentage of upstream catchment area.

```{r REC2}
df <- data.frame(df)
df[,57:85] <- (df[,57:85] / df[,34])*100 # Land cover data base
df[,86:97] <- (df[,86:97] / df[,34])*100 # Top rock SI database

```

Make rownames of dataframe UID and add TICI values to sampledata dataframe df.

```{r df_rownames}
row.names(df) <- df$UID
TICI <- data.frame(TICI)
row.names(TICI) <- TICI$UID

df <- merge(df, TICI[,3:7], by='row.names', all=TRUE)
df$Row.names <- NULL
colnames(df)[6] <- "X.SampleID"
df <- relocate(df, "X.SampleID")
row.names(df) <- df$X.SampleID
```

Add to a sample_data phyloseq object and add an 'X' to sample names

```{r sample_to_phyloseq}
sampledata <- sample_data(df)
sample_names(sampledata) <- paste0("X", sample_names(sampledata))
```

## Taxa read counts

This section prepares the taxa read count data for input into phyloseq Taxa IDs are provided by their NCBI ID. Samples are columns and taxa rows. Values represent the total read counts for that particular taxa for each sample.

```{r taxa}
# Select relevant columns from the Wilderlab output
otumat <- data.frame(taxa_data[c(3,6:ncol(taxa_data))])

# Change TaxID to match ncbi database
otumat[which(otumat$TaxID == 10000005),1] = 27462 # Austrosimulium changed to genus
otumat[which(otumat$TaxID == 10000018),1] = 126351 # Changed to Gal spD
otumat[which(otumat$TaxID == 10000020),1] = 89553 # Changed to Giant Kokopu
otumat[which(otumat$TaxID == 10000033),1] = 309669 # Changed to genus
otumat[which(otumat$TaxID == 10000037),1] = 65076 # Changed to genus
otumat[which(otumat$TaxID == 10000038),1] = 226931 # Changed to Common Bully
otumat[which(otumat$TaxID == 10000043),1] = 98305 #Changed to genus
otumat[which(otumat$TaxID == 10000052),1] = 75837 # Changed to Brown Teal
otumat[which(otumat$TaxID == 10000053),1] = 126303 # Changed to Gal dep
otumat[which(otumat$TaxID == 10000054),1] = 126303 # Changed to Gal dep

#Combine rows with same TaxID
otumat <- aggregate(x = otumat, by = list(otumat$TaxID), FUN = function(x) na.omit(x)[1])[,-1]

# Now import into phyloseq
row.names(otumat) <- otumat$TaxID
taxdf <- otumat$TaxID # data required for obtaining taxonomic information
otumat$TaxID <- NULL
otumat <- as.matrix(otumat)
OTU = otu_table(otumat, taxa_are_rows = TRUE)
```

## Taxonomy

Here, the NCBI database is called to obtain the taxonomic information for each taxa present in the data set. This data includes the taxonomic ranks for each taxa, such as Super Kingdom, Kingdom, Phylum, Class, Order, Family, Genus and Species. The package 'metacoder' is used here to connect to NCBI. It is useful to use 'write.csv' to produce a list which can be filtered in excel to easily explore the results.

```{r taxonomy}
# Use library(taxa and metacoder) to do taxonomy for phyloseq
x <- lookup_tax_data(taxdf, type = "taxon_id") #takes a while to run
y <- taxonomy_table(x, use_ranks = c("superkingdom", "kingdom", "phylum", "class", "order",
        "family", "genus", "species"), add_id_col = TRUE)
y <- data.frame(y)
rownames(y) <- y$taxon_id
y$taxon_id <- NULL
taxmat <- as.matrix(y)

# write.csv(taxmat, "tax_table.csv")

# Make a phyloseq object
TAX = tax_table(taxmat)
```

## Build phyloseq object

The final step is to combine the taxa abundance data with the sample data (metadata) and taxonomic information. This data is all saved as a phyloseq object called 'physeq'. This is a useful format to save the data as it incorporates all the sequencing information, metadata and taxonomic information.

```{r phyloseq_object}
physeq = phyloseq(OTU, TAX, sampledata) # Sample data not matching UID
physeq
#saveRDS(physeq, file = "April 2022 data.rds")
```

# Superkingdom: Bacteria

The following details an exploratory data analysis and some detail on the functional ecology of the bacterial populations detected during this survey.

First we must filter the 'physeq' data set to only bacterial taxa.

```{r}
#Filter to only bacteria
bac <- subset_taxa(physeq, superkingdom == "Bacteria")
#remove 'kingdom' rank
tax_table(bac) <- tax_table(bac)[,c(1,3:8)]
```

First, we want to understand the distribution of read counts. These results are very unusual as most sequences are common throughout the data set.

```{r}
sort(sample_sums(bac))
hist(sample_sums(bac), xlab="Total Read Counts")

```

Here we describe the most abundant bacteria taxa. A total of 562,621 bacterial reads were detected. Taxa ID 237 (Genus Flavobacterium) was the most abundant with 212,242 reads, followed by 101 (Genus Flectobacillus) with 41,999 reads.

```{r}
sum(taxa_sums(bac))
sort(taxa_sums(bac))[(length(taxa_sums(bac))-10):length(taxa_sums(bac))]

```

Of those reads, the phyla Proteobacteria and Bacteroidetes are the most diverse with 164 and 57 taxa, respectively.

```{r ntaxa}
ntaxa(subset_taxa(bac, phylum == "Proteobacteria"))
ntaxa(subset_taxa(bac, phylum == "Bacteroidetes"))
```
Taxonomic tree using the `metacoder` package.

```{r bac_tree}
heat_tree(parse_phyloseq(bac),
          node_size = n_obs,
          node_color = n_obs,
          node_label = taxon_names,
          tree_label = taxon_names,
          layout = "davidson-harel")

```

To understand the taxonomic composition of each site, we will rarefy the dataset to an even sampling depth so that we can plot a comparable bar plot.

The sampling size is determined as 90% of the sample with the smallest sampling size (approximately 2700). Data for each Site is combined to aid interpretation.

The samples were largely dominated by taxa belonging to either Bacteroidetes or Proteobacteria. The relative proportions of these two phyla varied between sites.

```{r}
bac.site <- merge_samples2(bac, "Site", fun_otu = sum, funs = list(), reorder = FALSE)
bac.site.rel <- transform_sample_counts(bac.site, function(x) x/sum(x)) # Transform counts to relative proportions
bac.rarefied = rarefy_even_depth(bac.site, rngseed=1, sample.size=0.9*min(sample_sums(bac.site)), replace=F)
plot_bar(bac.rarefied, x="Site", fill="phylum") + facet_wrap(~SubCatch, scales = "free_x", nrow=3)
```

There was a general trend of bacterial richness increasing from upstream to downstream sites.

```{r}
plot_richness(bac.rarefied, x="Downstream.distance", color="Type", measures=c("Observed", "Chao1", "Shannon")) + 
  geom_smooth(method='lm', se=TRUE)
plot_richness(bac.rarefied, x="Temp", measures=c("Observed", "Chao1", "Shannon")) + 
  geom_smooth(method='lm', se=TRUE)
plot_richness(bac.rarefied, x="Altitude", measures=c("Observed", "Chao1", "Shannon")) + 
  geom_smooth(method='lm', se=TRUE)
plot_richness(bac.rarefied, x="us_catarea", measures=c("Observed", "Chao1", "Shannon")) + 
  geom_smooth(method='lm', se=TRUE)

```

PCoA of entire bacterial dataset. Probably best to do at a site level to have an interpretable MDS plot

``` {r}
bac.ord <- ordinate(bac, "PCoA", "bray")
p1 = plot_ordination(bac, bac.ord, type="taxa", color="phylum", title="Bacterial phyla")
p1
p2 = plot_ordination(bac, bac.ord, type="samples", color="Site") 
p2 + geom_polygon(aes(fill=Site)) + geom_point(size=5) + ggtitle("samples")
p2
```

The first two axes of PCoA using Bray Curtis distances explained 38.6% of the variation in bacterial community composition. There were no clearly discernible patterns between the sites and bacterial community composition. No clear patterns emerged between tributary and main stem community composition. There was some evidence of weak clustering occurring between sites at similar upstream distances. For instance, T1, Waipouri 1, Lee Creek and Owhiro Stream all had similar compositions.

``` {r}
bac.ord <- ordinate(bac.site, "PCoA", "bray")
p2 <- plot_ordination(bac.site, bac.ord, type="samples", color="Downstream.distance", shape="Type") +
  theme_bw() +
  geom_point(size=10) + 
  geom_text(aes(label = Site), size = 5, vjust = 0.5, hjust = 0.5, color="black") +
  scale_colour_gradient(low = "yellow", high = "darkblue") +
  ggtitle("PCoA: Bray-Curtis")
p2

```

A network representation of samples, representing samples that occur in similar profiles of bacterial community.

```{r}
ig = make_network(bac, type = "samples", distance = "bray", max.dist = 0.3)
plot_network(ig, bac, color = "Site", shape = "SubCatch", line_weight = 0.4) #, label = NULL)
```

A network representation of OTUs, representing communities of bacteria that occur in similar profiles of samples

```{r}
# prune to just the top 100 most abundant OTUs across all samples (crude).
jg = make_network(bac, "taxa", "bray", 0.3)
plot_network(jg, bac, "taxa", color = "phylum", line_weight = 0.4) #, label = NULL)
```

## Bacterial environemntal relationship

First need to get data into meco format. Including Archae for this analysis

```{r meco_input}
#First remove kingdom rank and rename superkingdom to kingdom
bac2 <-  merge_phyloseq(subset_taxa(physeq, superkingdom=="Bacteria"), subset_taxa(physeq, superkingdom=="Archaea"))
bac2 <- merge_samples2(bac2, "Site", fun_otu = sum, funs = list(), reorder = FALSE)

tax_table(bac2) <- tax_table(bac2)[,c(1,3:8)]
colnames(tax_table(bac2)) <- c("Kingdom", "Phylum", "Class", "Order", "Family",  "Genus", "Species")#Transfer to meco
#bac2 <- tax_glom(bac2, taxrank = "Genus")
meco_dataset <- phyloseq2meco(bac2)
```

Create trans_env object for downstream analysis. Also build abundance table.

```{r add_env_data}
t1 <- trans_env$new(dataset = meco_dataset, env_cols = c(13,17,19,22,57,64,69:72,76:78,81,84,85)) #Conductivity not included due to not being recorded at all sites
t1$dataset$cal_abund()
```

Use 'cal_diff' to test the significance of variables between groups

```{r cal_diff}
# use Wilcoxon Rank Sum and Signed Rank Tests as an example
t1$cal_diff(group = "SubCatch", method = "wilcox")
t1$res_diff[, c(1, 2, 4, 6, 7)]
```

Present the results as box plots

```{r env_boxplots}
t1$cal_diff(method = "anova", group = "SubCatch")
# place all the plots into a list
tmp <- list()
for(i in colnames(t1$data_env)){
    tmp[[i]] <- t1$plot_diff(measure = i, add_sig_text_size = 3, xtext_size = 8, ytext_size = 5) + theme(plot.margin = unit(c(0.1, 0, 0, 1), "cm"))
}
plot(arrangeGrob(grobs = tmp, ncol = 4))
```

Now show autocorrelation between variables

```{r env_autocorrelation}
# require GGally package installed
t1$cal_autocor()
```

And the distribution of variables and autocorrelation between the different groups.

```{r env_group_correlation}
t1$cal_autocor(group = "SubCatch")
```

Now use db-RDA to summarise data graphically with 10 most common families

```{r dbRDA}
# use Genus
t1$cal_ordination(method = "RDA", taxa_level = "Family")
# As the main results of RDA are related with the projection and angles between different arrows,
# we adjust the length of the arrow to show them clearly using several parameters.
t1$trans_ordination(show_taxa = 10, adjust_arrow_length = TRUE, max_perc_env = 1.5, max_perc_tax = 1.5, min_perc_env = 0.2, min_perc_tax = 0.2)
# t1$res_rda_trans is the transformed result for plotting
t1$plot_ordination(plot_color = "SubCatch")
```

Use Mantel test to understand relationship of bacterial communities with the env_data. Downstream distance the only important factor.

```{r mantel}
t1$cal_mantel(use_measure = "bray")
# return t1$res_mantel
t1$res_mantel
```
Now look at specifi bacterial families and their relationships with the env_data
```{r cor_env}
t1$cal_cor(use_data = "Family", p_adjust_method = "fdr")
# default ggplot2 method with clustering
# filter Family that do not have at least one *** and add clusters
t1$plot_cor(pheatmap = TRUE, filter_feature = c("", "*"), color_palette = rev(RColorBrewer::brewer.pal(n = 9, name = "RdYlBu")))
```

## Bacteria functional traits

Functional traits are analysed using the 'meco' package. Need to build a new meco 'trans_fun' object called t2.

```{r assign_functions}
# create object of trans_func
t2 <- trans_func$new(meco_dataset)
# mapping the taxonomy to the database
# this can recognize prokaryotes or fungi automatically if the names of taxonomic levels are standard.
# for fungi example, see https://chiliubio.github.io/microeco_tutorial/other-dataset.html#fungi-data
# default database for prokaryotes is FAPROTAX database
t2$cal_spe_func(prok_database = "FAPROTAX")
# return t2$res_spe_func, 1 represent trait exists, 0 represent no or cannot confirmed.
t2$res_spe_func[20:25, 1:10]
```

The percentages of OTUs having the trait reflect functional redundancy in the community.

```{r fun_relative}
# calculate the percentages for communities
# here do not consider the abundance
t2$cal_spe_func_perc(abundance_weighted = FALSE)
## The result table is stored in object$res_spe_func_perc ...
t2$res_spe_func_perc[20:25, 1:10]
```

```{r fun_network}
# construct a network for the example
network <- trans_network$new(dataset = t2, cal_cor = "base", taxa_level = "OTU", filter_thres = 0.0001, cor_method = "spearman")
network$cal_network(p_thres = 0.01, COR_cut = 0.7)
network$cal_module()
# convert module info to microtable object
meco_module <- network$trans_comm(use_col = "module")
meco_module_func <- trans_func$new(meco_module)
meco_module_func$cal_spe_func(prok_database = "FAPROTAX")
meco_module_func$cal_spe_func_perc(abundance_weighted = FALSE)
meco_module_func$plot_spe_func_perc()
# M represents module, ordered by the nodes number from high to low
```

```{r functional_groups}
# If you want to change the group list, reset the list t2$func_group_list
t2$func_group_list
# use show_prok_func to see the detailed information of prokaryotic traits
t2$show_prok_func("nitrate_reduction")
```

```{r}
# then we try to correlate the res_spe_func_perc of communities to environmental variables
t3 <- trans_env$new(dataset = meco_dataset, add_data = data.frame(sample_data(bac2)[,c(13,17,19,22,57,64,69:72,76:78,81,84,85)]))
t3$cal_cor(add_abund_table = t2$res_spe_func_perc, cor_method = "spearman")
t3$plot_cor(pheatmap = TRUE)
```

```{r}
m1 <- trans_abund$new(dataset = meco_dataset, taxrank = 'Phylum', groupmean = "Site", ntaxa = 5)
m1$plot_pie(facet_nrow = 5)
```

## Bacteria differentiation between three catchment areas

```{r}
meco_dataset2 <- meco_dataset
meco_dataset2$tax_table %<>% tidy_taxonomy
meco_dataset2$tidy_dataset()
meco_dataset2$cal_abund()
# metastat analysis at Genus level
t4 <- trans_diff$new(dataset = meco_dataset2, method = "metastat", group = "SubCatch", taxa_level = "Family")
```

```{r}
# select_group should be one of groups in t1$res_diff$Comparison
t4$plot_diff_abund(use_number = 1:20, select_group = "Plains - Maniatoto")
```


```{r}
t4 <- trans_diff$new(dataset = meco_dataset2, method = "lefse", group = "SubCatch", alpha = 0.5, lefse_subgroup = NULL)
# see t1$res_diff for the result
# From v0.8.0, threshold is used for the LDA score selection.
t4$plot_diff_bar(threshold = 4)
# we show 20 taxa with the highest LDA (log10)
t4$plot_diff_bar(use_number = 1:30, width = 0.8, group_order = c("Plains", "Strath", "Maniatoto"))
```

```{r}
# clade_label_level 5 represent phylum level in this analysis
# require ggtree package
t4$plot_diff_cladogram(use_taxa_num = 100, use_feature_num = 50, clade_label_level = 5, group_order = c("Plains", "Strath", "Maniatoto"))
```

```{r}
# use Genus level for parameter taxa_level, if you want to use all taxa, change to "all"
# nresam = 1 and boots = 1 represent no bootstrapping and use all samples directly
t4 <- trans_diff$new(dataset = meco_dataset2, method = "rf", group = "SubCatch", taxa_level = "Family")
# plot the MeanDecreaseGini bar
g1 <- t4$plot_diff_bar(use_number = 1:20, group_order = c("Plains", "Strath", "Maniatoto"))
# plot the abundance using same taxa in g1
g2 <- t4$plot_diff_abund(group_order = c("Plains", "Strath", "Maniatoto"), select_taxa = t4$plot_diff_bar_taxa)
# now the y axis in g1 and g2 is same, so we can merge them
# remove g1 legend; remove g2 y axis text and ticks
g1 <- g1 + theme(legend.position = "none")
g2 <- g2 + theme(axis.text.y = element_blank(), axis.ticks.y = element_blank())
gridExtra::grid.arrange(g1, g2, ncol = 2, nrow = 1, widths = c(2, 1.7))
```

```{r}
t4 <- trans_diff$new(dataset = meco_dataset2, method = "wilcox", group = "SubCatch", taxa_level = "Family", filter_thres = 0.05)
# filter something not needed to show
t4$res_diff %<>% subset(Significance %in% "***")
t4$plot_diff_abund(use_number = 1:20, add_sig = T, add_sig_label = "Significance")
```

```{r}
t4 <- trans_diff$new(dataset = meco_dataset2, method = "anova", group = "SubCatch", taxa_level = "Genus", filter_thres = 0.05)
t4$plot_diff_abund(use_number = 1:10, add_sig = T, coord_flip = F)
```

```{r}

```